branch.go

```diff
@@ -0,0 +1,196 @@
+type IfBranch[T Steper] struct {
+	Target      T // the target to check
+	BranchCheck BranchCheck[T]
+	ThenStep    []Steper
+	ElseStep    []Steper
+	Cond        Condition // Cond is the When condition for both ThenStep and ElseStep, not target Step!
+}
+
+// Then adds steps to the Then branch.
+func (i *IfBranch[T]) Then(th ...Steper) *IfBranch[T] {
+	i.ThenStep = append(i.ThenStep, th...)
+	return i
+}
+
+// Else adds steps to the Else branch.
+func (i *IfBranch[T]) Else(el ...Steper) *IfBranch[T] {
+	i.ElseStep = append(i.ElseStep, el...)
+	return i
+}
+
+// When adds a condition to both Then and Else steps, not the Target!
+// Default to DefaultCondition.
+func (i *IfBranch[T]) When(cond Condition) *IfBranch[T] {
+	i.Cond = cond
+	return i
+}
+func (i *IfBranch[T]) isThen(isThen bool) func(ctx context.Context, ups map[Steper]StatusError) StepStatus {
+	return func(ctx context.Context, ups map[Steper]StatusError) StepStatus {
+		if status := ConditionOrDefault(i.Cond)(ctx, ups); status != Running {
+			return status
+		}
+		if i.BranchCheck.OK == isThen {
+			return Running
+		}
+		return Skipped
+	}
+}
+func (i *IfBranch[T]) Done() map[Steper]*StepConfig {
+	return Steps().Merge(
+		Steps(i.Target).AfterStep(func(ctx context.Context, s Steper, err error) error {
+			i.BranchCheck.Do(ctx, i.Target)
+			return err
+		}),
+		Steps(i.ThenStep...).When(i.isThen(true)),
+		Steps(i.ElseStep...).When(i.isThen(false)),
+		Steps(append(append([]Steper{},
+			i.ThenStep...), i.ElseStep...,
+		)...).
+			DependsOn(i.Target).
+			BeforeStep(func(ctx context.Context, s Steper) (context.Context, error) {
+				if i.BranchCheck.Error != nil {
+					return ctx, i.BranchCheck.Error
+				}
+				return ctx, nil
+			}),
+	).Done()
+}
+
+// Switch adds a switch branch to the workflow.
+//
+//	Switch(someStep).
+//		Case(case1, func(ctx context.Context, someStep *SomeStep) (bool, error) {
+//			// branch condition here, true to select this branch
+//			// error will fail the case
+//		}).
+//		Default(defaultStep), // the step to run if all case checks return false
+//	)
+func Switch[T Steper](target T) *SwitchBranch[T] {
+	return &SwitchBranch[T]{Target: target, CasesToCheck: make(map[Steper]*BranchCheck[T])}
+}
+
+// SwitchBranch adds target step, cases and default step to workflow,
+// and check the target step and determine which branch to go.
+type SwitchBranch[T Steper] struct {
+	Target       T
+	CasesToCheck map[Steper]*BranchCheck[T]
+	DefaultStep  []Steper
+	Cond         Condition
+
+	wg sync.WaitGroup // default need to wait all cases to finish check
+}
+
+// BranchCheck represents a branch to be checked.
+type BranchCheck[T Steper] struct {
+	Check BranchCheckFunc[T]
+	OK    bool
+	Error error
+}
+
+func (bc *BranchCheck[T]) Do(ctx context.Context, target T) {
+	bc.OK, bc.Error = bc.Check(ctx, target)
+}
+
+// Case adds a case to the switch branch.
+func (s *SwitchBranch[T]) Case(step Steper, check BranchCheckFunc[T]) *SwitchBranch[T] {
```

Comment:
How about turning the order of `step` and `check`? I think that is more natural.

Reply:
Actually, putting `step` at the first position is intended.
Because all functions accepting a `step` and being added into workflow have this patter.
```
flow.Step(step),
flow.Name(step, "name"),
flow.If(step, ...).
    Then(thenStep).
    Else(elseStep),
flow.Switch(step).
    Case(case1, ....).
    Default(defaultStep),
```